plugins/catalog-react/src/components/EntityRefLink/EntityRefLinks.tsx

```diff
@@ -35,19 +35,29 @@ export type EntityRefLinksProps = {
  * @public
  */
 export function EntityRefLinks(props: EntityRefLinksProps) {
-  const { entityRefs, defaultKind, ...linkProps } = props;
+  const { entityRefs, defaultKind, getTitle, ...linkProps } = props;
   return (
     <>
-      {entityRefs.map((r, i) => (
-        <React.Fragment key={i}>
-          {i > 0 && ', '}
-          <EntityRefLink
-            {...linkProps}
-            entityRef={r}
-            defaultKind={defaultKind}
-          />
-        </React.Fragment>
-      ))}
+      {entityRefs.map((r, i) => {
+        const isCompoundEntityRef =
+          getTitle && typeof r !== 'string' && !('metadata' in r);
+
+        const title = isCompoundEntityRef
+          ? getTitle(r as CompoundEntityRef)
```

Comment:
We'll want to parse the ref if it's a string ref, and convert to to a compound ref if it's an entity. `@backstage/catalog-model` has helpers for both of those, although not a single function that handles all the cases :<

Btw, can I ask what your use-case for this addition is? I've realized that `EntityRefLinks` itself is very slim so it may make sense to use `EntityRefLink` directly instead.

Reply:
Ok, I think the part with conversion would be not hard to handle.

Use case is to show all team names related to a domain in the company. 
Of course I can do it just in our project without upstreaming the changes.

But I have a question to you, how do you see it, with `EntityRefLink` and `EntityRefLinks`. 
Usually names are `internal` to the system, and what is shown on the screen to the user mostly humanised. 
At this component to show humanised text is more of exception than default behaviour. 
I understand that implementation is easier at this case, but such places goes first to customise for users who cares about UI/UX.

Second aspect to it, I noticed that some developers looking at `Backstage` code as an example how it should be done. So if `EntityRefLinks` exists, maybe worths to make it more flexible for usage so there is no need to replace it with home-grown solution?

Reply:
Yep totally, I'm mainly trying to be a bit careful here so that we don't implement a feature that in practice isn't very useful. I'm thinking that especially if you know you have a full list of entities, then that's what you'd want to have in the callback. That way you can pick out things like display names and what not, which is a much better use-case for something like this.

What does your `getTitle` look like approximately? all of the logic involved

One way that I think we could make this shippable real fast is if we can figure out how to add a type parameter that forwards the types  of the entities being passed in. For example, if you pass in `['a', 'b', {apiVersion: ..., kind: ..., metadata: ...}]` as `entityRefs`, then `getTitle` would have the signature `(entity: string | Entity) => string | undefined`

Reply:
Sorry @Rugvip was busy, couldn't reply on your questions.

In my case I read explicitly defined field 'displayName' in Entity from all entities 

![Screenshot 2022-09-23 at 09 39 01](https://user-images.githubusercontent.com/2265094/191912783-1f5332e2-8e46-4694-9ab8-7d922f72cfe6.png)

![Screenshot 2022-09-23 at 09 39 07](https://user-images.githubusercontent.com/2265094/191912794-8c4ff4c4-14b3-4de8-b6b2-3e38cd7464ee.png)

So if to go one step forward, the ideal situation for the user using it is to specify the path, where to read the displayName in `jq` like style: ['.spec.profile.displayName'] and depends on the provided type by user, we can or fetch entities by `refs` or using entities as they are, to fill in titles. 

That would be the ideal situation

Reply:
Hmm, feeling `getTitle` is nicer than a `jq`-like thing if I'm gonna be honest. It's not really that much more verbose, and it lets you easily implement fallbacks etc.

Either way, looks like what you really want to do here is something like this (pseudo-ish, and added type annotations for clarity):

```tsx
const relatedGroups: GroupEntity[] = await fetchRelatedGroupsOf(entity);

return (
  <EntityRefLinks
    defaultKind="group"
    entityRefs={relatedGroups}
    getTitle={(e: GroupEntity) => e.spec.profile?.displayName}
  />
);
```

The automatic fetching of entities is a nice idea though, we could do something like this?

```tsx
return (
  <EntityRefLinks
    fetchEntities
    defaultKind="group"
    entityRefs={parentRelations}
    getTitle={(e: Entity) => (e as GroupEntity).spec.profile?.displayName}
  />
);
```

With that you'd end up with a guarantee that `getTitle` always receives a full entity. We'd only fill in the gaps by fetching the full entity of any entity refs that are passed through, full entities would be forwarded as they are.

What do you think? It's a bit of TypeScript wizardry and complicates `EntityRefLinks` quite a bit, but it does seem like a more powerful solution that solves a pretty common problem.

Reply:
Indeed, here is a question where is better to keep the complexity, inside the Backstage or on a user side. 
I can understand your point as a Backstage maintainer to keep the less code as possible to maintain. 
It is like 2 sides of the coin. 
From user perspective, the more components/APIs which platform provides to develop faster makes the feeling of the platform as more mature. 
If to look at Spring community, especially with emerging of Spring Bootstrap, why it became so popular? Cause as a user of this platform I have to configure few things and the whole world is configured for me 😃 

This is a good example for this component that making a so slight change requires to produce so much code on my (user) side. 

If you prefer `getTitle` style it's perfectly fine for me, this way it gives to the end user more room to do even more customization if needed, not only read the field as with `jq`, but also to apply some transformation. 

Regarding the fetching of entities inside. Do you want to have an explicit field for that like `fetchEntities=true` or do it implicitly? As we can go even once step forward in that. Instead of providing the list of entity refs in `entityRefs={parentRelations}`, we can define the root entity and defining `relationType`. The call of 

```
const relations = getEntityRelations(entity, relationType, {
    kind: 'group',
  });
```

can be inside. And we can then also add some `mapReduce` field as a column for a transformation. 

I checked the usage of this component inside the Backstage itself, and i see this pattern everywhere.

Reply:
Regarding complexity I only pointed out the complication of the component from the point of view that you would have to implement that and it's more work. 😅 Completely fine with having plenty of complexity within the codebase, the only trouble is if the interfaces leak that complexity.

I do actually think that there are enough use-cases where fetching entities isn't required and where you want to display some other list of entity links than specifically relations. It also composes pretty well so not too worried about the extra steps needed tbh. That is, yes, I think it should be an explicit field to make fetching happen, and that it's enough to just pass in entity refs without any logic around relations.

Reply:
Ah ok ok, indeed, always should be other way around.

Cool, that already would be a good step forward, especially if you could think of many cases to keep the same interface.

The only question what i have is to the interface of `getTitle` when we toggle `fetchEntities`.

If `fetchEntities=false`, and user provides the list of strings to `entityRefs` how can we convert it to list of Entities?
For me it's a bit contradictory. What am I missing here? 🙄 😅 

For me it looks more like implicit, if list of entities is provided, we do nothing, if list of refs or strings, we do auto-conversion. 

Or you mean that if user provides a `list of strings` in `entityRefs` and `fetchEntities=false`, `getTitle` will get as input a `string`?

Reply:
I'm thinking something like this:

```ts
type EntityRefLinksProps<TRef extends string | CompoundEntityRef | Entity> = {
  fetchEntities?: false,
  entityRefs: TRef[],
  getTitle?(entity: TRef): string | undefined
} | {
  fetchEntities: true,
  entityRefs: TRef[],
  getTitle?(entity: TRef extends Entity ? TRef : Entity): string | undefined
}
```

With that we'll forward the refs as whatever type they are if `fetchEntites` isn't set, that way it's up to the caller to either handle any possible entity formats in the `getTitle` callback, or to normalize the entity refs upfront.

In the case where `fetchEntities` is set, we'll forward any full entities as is, and that'll be visible in `getTitle`, but any entity refs will be fetched and converted into an `Entity` for `getTitle`